### ISSUE TITLE
Add p-value or f-stat?

### ISSUE THREAD

**Issue by moptis** [dswah/pyGAM#163]
Hello, 

Is there a plan to output p-values or f-stats to the individual predictors?

Thanks!

**Comment by dswah** [comment]
Hi @moptis . Yes! the plan is to add these asap!

Also, I happily welcome collaborators. :wink: 
Do you have experience with these tests?

**Comment by dswah** [comment]
hi @moptis i've just added p-values to the model summary!

you should be able to see the following for a fitted GAM.
![image](https://user-images.githubusercontent.com/11619412/38405194-82424870-3923-11e8-9ed6-17c23d6ebb7f.png)

plz let me know how this works for you.

**Comment by dswah** [comment]
@moptis 😒 that does seem weird. also strange that the intercept is the same. it looks like all the pvalues are getting as small as possible... im going to dig in a little more.

what happens when you dont normalize y?

and could i ask you to add some noise (maybe a lot?) to one of the features and show me the pvalues again?

lets fix this.

**Comment by moptis** [comment]
@dswah if I don't normalize y, all the p-values are equal to 1. If i use different combinations of features (e.g. x1, x2, x3; x1, x2; x1; etc.) I get the same result: the same low p-value for all features if 'y' is normalized and 1 if 'y' is not normalized.

I wonder if I'm using some package dependency that is out of date?

**Comment by moptis** [comment]
Also for reference is a snapshot of the result in R using the 'mgcv' library. 

![image](https://user-images.githubusercontent.com/32751681/38999573-5ebb71c4-43af-11e8-8816-a53df0bca42f.png)

**Comment by dswah** [comment]
@moptis thanks for this info. it is really helpful.

I was able to reproduce your findings.

When i divide `y` by its standard deviation, then all my pvalues become tiny and equal, just like yours. 
When i multiply `y` by a large number, then i get all the pvalues are equal to `1`.

I hope this is some obvious bug in the math.

**Comment by dswah** [comment]
there is at least one obvious bug in the math :)
https://github.com/dswah/pyGAM/pull/171

**Comment by dswah** [comment]
@moptis I've pushed a new version of the code that fixes a bug in the F-statistic (used in the p-value).

could you update the library and show me the results again?

**Comment by moptis** [comment]
@dswah Sorry for the delay. I was out of office over the last 10 days. I'll update the library soon and will reach out when I have results.

**Comment by moptis** [comment]
@dswah I'm still getting the same result with version 0.5.2:

![image](https://user-images.githubusercontent.com/32751681/40240808-aa1786ca-5a76-11e8-97f3-1eb238d0f7a8.png)

**Comment by dswah** [comment]
@moptis wow Im so sorry i havent answered yet!
let me take another look!

**Comment by dswah** [comment]
@moptis grrr... hmm the p-values certainly look wrong, but are they still scaling wrt the data scale?

**Comment by dswah** [comment]
@alexgain i believe it :/
I frequently see these values as well.

I'm not sure what to do until we fix this bug.
- Should we remove the p-values from the model summary?
- should we leave them in, but add a warning that there is a known bug in the computation?
- something else?

your ideas are very welcome

**Comment by moptis** [comment]
@dswah @alexgain I'm not so sure these are bugs. I think 1e-16 is just the lowest possible value. I see the same in the R mgcv package, except the value is 2e-16. So if I see 1e-16, I just know that the feature is VERY statistically significant.

**Comment by canthonyscott** [comment]
Hi all, Is there any new information regarding this? I am about to perform a large analysis and would love to use pyGAM but I do need the p-values for each feature in the model. Are they trust-worthy? Is it save to move forward?

Thanks for the guidance!

**Comment by shyamcody** [comment]
@dswah probably a reason the f-statistics are shooting higher values is that maybe the wald statistics being calculated is wrong according to [this cross-validated answer](https://stats.stackexchange.com/questions/159443/gam-summary-f-values). To summarize, the wald statistics we calculate is beta* inv(V)*beta_transpose. But with rank r in penalized model, Simon wood used beta*((V)^-rank)*beta which is lesser value in case of rank>1; and therefore will return higher p-value as the corresponding CDF will be less.

**Comment by dswah** [comment]
@shyamcody nice find. I took a look at the cross-validated answer, and it looks like we [accounted for the rank of the covariance matrix in the F-statistic](https://github.com/dswah/pyGAM/blob/b57b4cf8783a90976031e1857e748ca3e6ec650b/pygam/pygam.py#L1261).

(also, I believe we are correctly following the updated formula from the [errata](https://people.maths.bris.ac.uk/~sw15190/igam/iGAMerrata-12.pdf) as shown in equation for p.195.)

Then again, I am not sure what Wood means by "Vr- is the rank-r pseudoinverse of the covariance matrix." 
In the code I have simply calculated the pseudo-inverse and collected the rank returned by the scipy's `pinv`, which I pass to the corresponding statistic...

Perhaps like you suggest we need to somehow scale the pseudoinverse by the computed rank?
I am looking at [mgcv source](https://www.rdocumentation.org/packages/mgcv/versions/1.8-31/source) code to try to get an idea of how they compute their p-values, but I haven't understood much yet.

With respect to poor conditioning of the determinant matrix, let's look into this! 
Perhaps the un-centered features are causing problems in the covariance estimates?

**Comment by shyamcody** [comment]
hey @dswah, my bad in reading. I misunderstood the r-- notation as the power of (-r) from the StackOverflow answer. Can you exactly mention how are you creating the parameter covariance matrix or refer to where in Simon's book it is written? my idea about the covariance matrix calculation is not clear and therefore comments may be insignificant. And yes, I too believe the unpenalized model related implementation is correct.

**Comment by dswah** [comment]
@shyamcody here is the [seemingly innocent line](https://github.com/dswah/pyGAM/blob/b57b4cf8783a90976031e1857e748ca3e6ec650b/pygam/pygam.py#L1040) where the covariance matrix is computed.

The action happens between pg 184 and 185 of the [1st edition book](https://reseau-mexico.fr/sites/reseau-mexico.fr/files/igam.pdf).

**Comment by uditrana** [comment]
Has this been fixed? If the P scores are wrong than the significance codes are also extremely wrong, right?

**Comment by shyamcody** [comment]
The issue is that, me and dswah were not able to find out what exactly is
wrong in the code. Can you figure it out?

**Comment by uditrana** [comment]
I investigated slightly but unfortunately do not have the domain expertise nor the time to investigate. I was just experimenting lightly with GAMs anyways. Might have to switch to R and mgcv if I end up wanting to use a GAM for my model...

I also noticed that the ordering in which you passed in features seemed to change results (not sure if this is related or not).

**Comment by jonathan-taylor** [comment]
Hmm... this line is a bit curious -- why center here? https://github.com/dswah/pyGAM/blob/b57b4cf8783a90976031e1857e748ca3e6ec650b/pygam/pygam.py#L1259

**Comment by jonathan-taylor** [comment]
Another issue here: https://github.com/dswah/pyGAM/blob/b57b4cf8783a90976031e1857e748ca3e6ec650b/pygam/pygam.py#L1271

For an F distribution it should be score / rank instead of score -- the F_{n,d} (n:numerator, d:denominator) dbn is defined as (\chi^2_n/n)/(\chi^2_d/d)

**Comment by jonathan-taylor** [comment]
For models with linear terms only and `lam=0`, the p-values do agree with `statsmodels.OLS`.

**Comment by Sorkanius** [comment]
any plan for updating these issues? Looking for a solid python GAM library, such as R's mgcv

**Comment by freblisna** [comment]
With #370 fixed by #404, does that fix this issue too?

**Comment by dswah** [comment]
@freblisna Unfortunately they are not related!

the variance/std mixup just affected the calculation of likelihoods. 

the wrong p-values are due to pyGAM using the approximation from Simon Wood's old 2006 text on GAMs. 

this approximation is known to reject the null hypothesis to readily and has been updated in the 2014 text. 

now it just needs to be implemented in pyGAM!

**Comment by vsl366** [comment]
Hello,
I'm Vishal and I'm Interested in working on this issue, esp for GSOC ! I am new to this and while I wasnt aware much about GAMs, this feels completely intuitive, and also efficient as it just stacks Basis Functions for every feature to model a curve using linear equations. Ill just tell what I understood and what i feel like is supposed to change, Please let me know if Im right.

Every feature is represented as
f(x) = Beta_1*B₁(x) + Beta_2*B₂(x) + ...
and Expected value is represented as
g(E(y)) = intercept + f(x_1) + f(x_2) ...
Adding a penalty term S, and a weight W for each data point 
we get Beta= inv(XᵀWX + S) XᵀWz

For a feature n to be significant, f(x) should not be zero, thus Beta_j should not be zero for atleast one value of j
To test H0 : Beta_j = 0, we use Walds statistic
T = Beta_jᵀ * inv(VBeta) * Beta_j

Now the issue is, In Woods 2006, the probability used to compute p value is computed by
P( ChiSq(rank) > T)

On reading Woods 2014, what I understand that we must do is to calculate P value in terms of
P( e1*ChiSq(1) + e2*ChiSq(1) + enChiSq(dof - 2) where e stands for eigen value
And Also Compute T with the rank Truncated pseudoinverse, not the current one.

Since the distribution is not a standard one, what i read from woods was to either use the resource intensive Davies method, or just a very good approximation using Liu Tang Zang Moment approximation to get the p values.

Since this method considers even the partial degrees of freedom, im assuming p value increases, and the result becomes non significant, thus reducing rejections. Please let me know if Ive done any mistakes. As of now, i have only read small snippets of the Woods papers, did some background work. I havent checked the codebase much yet, but am eager to work upon this issue. Please let me know if anything is required, Thanks :)



TLDR : Compare the new T with e1*ChiSq(1) + e2*ChiSq(1) + ChiSq(deg of freedom -2) instead of ChiSq(rank), and use Liu et al for computing p value.

**Comment by Mudassiruddin7** [comment]
Finally got to this one. The problem was that _compute_p_value() was using the full algebraic rank of the Bayesian covariance submatrix as chi-squared degrees of freedom. For a penalized spline with 25 basis functions but EDoF ~3.5, the test was using df=25, which is massively overpowered. Replaced with Wood (2013b)'s rank-r truncated eigendecomposition: take r = ceil(EDoF), keep the top r eigenvectors of the covariance submatrix, project the coefficients onto that basis, test against chi-squared(r). Results are now consistent with mgcv. The KNOWN BUG comment in summary() is removed. Fix inPR. Reference: Wood (2013), Biometrika 100(1).

**Comment by freblisna** [comment]
Thanks so much!!!

**Comment by Kinara2020** [comment]
Hii @dswah ,
Myself Kinara Patel, and I am currently pursuing my B.Tech in Information and Communication Technology at Pandit Deendayal Energy University (PDEU), India. I am writing to express my interest in contributing to pyGAM under Google Summer of Code 2026.
After reviewing all the project ideas, I have chosen the following:
	∙	Primary: Modern P-Value Computation (Issue #163)
	∙	Secondary: Scikit-Learn Interface (Issue #422)
I have already cloned the repository and studied the current implementation of GAM._compute_p_values(). I can see it uses the Wood 2006 approximation, and I have started reading Wood (2014) to understand the improved Wald-like test approach and how it accounts for effective degrees of freedom more accurately.
My relevant skills include:
	∙	Python (intermediate), NumPy, SciPy, pandas
	∙	scikit-learn and general ML model building
	∙	Basic statistics including hypothesis testing and p-values
	∙	Hands-on experience building a real-time hand gesture recognition system using OpenCV in Python
I have a few questions to get started:
	1.	For Issue #163, is there a specific part of the implementation you would suggest I begin with first?
	2.	Are there any small beginner-friendly issues I can work on right now to better understand the codebase?
	3.	What is your preferred communication channel during the project?
I am committed to delivering quality work and look forward to contributing meaningfully to pyGAM.
Thank you for your time!
Best regards,
Kinara Patel
PDEU, India
GitHub: Kinara2020

**Comment by ChamRoshi** [comment]
> Finally got to this one. The problem was that _compute_p_value() was using the full algebraic rank of the Bayesian covariance submatrix as chi-squared degrees of freedom. For a penalized spline with 25 basis functions but EDoF ~3.5, the test was using df=25, which is massively overpowered. Replaced with Wood (2013b)'s rank-r truncated eigendecomposition: take r = ceil(EDoF), keep the top r eigenvectors of the covariance submatrix, project the coefficients onto that basis, test against chi-squared(r). Results are now consistent with mgcv. The KNOWN BUG comment in summary() is removed. Fix inPR. Reference: Wood (2013), Biometrika 100(1).

Cannot express how exciting this is to hear. Beside myself for the next release that includes this. The P values are a great feature of the library.
